Feature/smd 12 test specification processor

hbgwhite opened the pull request:
- Reorganized existing code
- Added Validator abstract class
     - Added Validator subclasses for min/max int/decimal
     - Added Validator for required DateTime/string/int/decimal
     - Added validator for max allowed string property length
- Added ValidatorCollection to simplify validation across multiple validation criteria
- Added Processor abstract class (implements IDisposable)
     - Added TestSpecificationProcessor
     - Added PropertyProcessor
     - Added IdentityProcessor
- Added IncorrectPackageTypeException

Processors use Validators to ensure that the XML attributes in their node of interest are correct. ValidatorCollection objects created against each field process multiple validations to determine correctness and produce error messages when things aren't correct. This pull request contains fully mapped and functional Processors for the TestSpecification, Property, and Identity XML nodes.

Error classes assigned as follows:
- Severe: This error will cause the system loader to break if you attempt to load the target XML into the database
- Degraded: This error violates the contract for this attribute as it exists in the documentation
- Benign: This error violates the documented contract for this attribute, but the attribute won't break the loader script and is not marked as a required field

hbgwhite committed:
Added common project to solution

hbgwhite committed:
Added TestSpecificationProcessor and IncorrectPackageTypeException. Completed validation for /testspecification

hbgwhite committed:
Validated testspecification sub-elements identifier and property. Created int and decimal version validation and string non-empty length validation. Downgraded .NET version to 4.5 to remain consistent with existing project.

hbgwhite committed:
Moved enums to their own folder

hbgwhite committed:
Moved common elements out of admin and into common

hbgwhite committed:
Added error writing and reporting convenience methods for TestSpecificationProcessor and subclasses

hbgwhite committed:
Centralized error handling

hbgwhite committed:
Added database validation subclass for TestSpecificationProcessor

hbgwhite committed:
Folder organization

hbgwhite committed:
Made sure writers are available

hbgwhite committed:
Quick fixes

hbgwhite committed:
Formatting, Processor stubs, Validators, Validator abstract class, Processor abstract class, Switched existing processors over to the new validators

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Tabulation/CsvWriter.cs:
```diff
@@ -22,49 +20,58 @@ public CsvWriter(TextWriter Writer)
                     // Do nothing
-                }
-                else if (value.IndexOfAny(sCsvSpecialChars) >= 0)
+                } else if (value.IndexOfAny(sCsvSpecialChars) >= 0)
```
Did ReSharper format like this?  Standard C# formatting is to have the `else` on it's own line and `{` or `}` on it's own line as well.  I see this same thing in a few places in this file.

hbgwhite replied:
it's a resharper thing. I'll modify my resharper profile to dump those on their own lines and re-run against the appropriate directories.

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Tabulation/ErrorHandling.cs:
```diff
@@ -0,0 +1,22 @@
+
+            var outMessage = string.Format(message, args);
+            writer.Write(new[] {testName, severity.ToString(), itemId, outMessage});
+#if DEBUG1
```
Is this directive supposed to be using `DEBUG1`?  Seems like a mistake.

Also, the `Debug.Fail` or `Debug.Assert` calls are already conditionally compiled, so wrapping them in `#if DEBUG` is not needed.

See: http://stackoverflow.com/questions/622743/are-debug-assert-debug-fail-automatically-conditionally-compiled-if-debug

hbgwhite replied:
This code was moved here from elsewhere in the project without modification. I'll remove the #DEBUG statements

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/MaxLengthValidator.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public override string GetMessage()
+        {
+            return $"[Length<={(Parameter is int ? Parameter.ToString() : string.Empty)}]";
```
The string interpolation is a C# 6 feature, I haven't thought thru if that is a problem or not, but we might need to validate that it's okay.

Also, I know if this was a traditional `string.Format()` you wouldn't need to call `.ToString()` on `Parameter`.  Is that the case with this string interpolation or not?

hbgwhite replied:
I took the ToString off. I'll run it through my integration tests and make sure everything is ok

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/MinDecimalValueValidator.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public override string GetMessage()
+        {
+            return "[MinDecimal]";
```
I'm guessing the Parameter should be included in the message

hbgwhite replied:
Adding parameters to all messages where they are present

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/MinIntValueValidator.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public override string GetMessage()
+        {
+            return "[MinInt]";
```
Include Parameter in the message string

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/RequiredDateTimeValidator.cs:
```diff
@@ -0,0 +1,23 @@
+        {
+            DateTime dateTimeValue;
+            var valueString = value as string;
+            return DateTime.TryParse(valueString, out dateTimeValue);
```
Up to you, but `return DateTime.TryParse(value as string, out dateTimeValue);` seems slightly cleaner

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/RequiredDecimalValidator.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            decimal decimalValue;
+            var valueString = value as string;
+            return decimal.TryParse(valueString, out decimalValue);
```
Same comment as above

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/RequiredIntValidator.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            int intValue;
+            var valueString = value as string;
+            return int.TryParse(valueString, out intValue);
```
ditto

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/MinIntValueValidator.cs:
```diff
@@ -0,0 +1,28 @@
+            int intParameter;
+            var valueString = value as string;
+            var parameterString = Parameter as string;
+            return !string.IsNullOrEmpty(valueString)
```
Devs have strong opinions on this.  Personally I prefer `String.IsNullOrEmpty()` and `Int32.TryParse` (the type name instead of the alias) when calling static methods and the alias (lower case) when using it as the variable.

I'm fine keeping it this way if the existing code used that as a standard.  You let me know.

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Loader/TestSpecificationErrorProcessor.cs:
```diff
@@ -0,0 +1,36 @@
+
+            AdminPackageUtility.ReportLoadError(Navigator.NamespaceURI, sXp_Publisher.Expression, validators.ObjectValidationErrors(Publisher));
+            // If it fails loader enforcement we're going to check if it meets spec enforcement.
+            base.IsValidPublisher();
```
Since the `ReportLoadError` is called before checking the Test spec IsValid.  Won't this add 2 errors for this one check?  Shouldn't the call to `ReportLoadError` here be contingent on what is returned from `base.IsValidPublisher`?

hbgwhite replied:
Since the database loader script and the spec errors aren't working off of the same value here, I wanted it to be clear if it violated both sets of validations or just one so I'm running both. The ErrorProcessor will produce severe error versions and the regular processor will produce degraded ones since this is a required field.

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/Validator.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public abstract class Validator
+    {
+        protected Validator(ErrorSeverity errorSeverity, object parameter)
```
I see that some Validators like `RequiredStringValidator` pass in `null` since they don't have anything relevant.  How about we make that parameter optional so it doesn't have to be initialized with null?

`protected Validator(ErrorSeverity errorSeverity, object parameter = null)`

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/ValidatorCollection.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace TabulateSmarterTestAdminPackage.Common.Validators
+{
+    public class ValidatorCollection
```
I think we can clean this up a bit.

First, let's make the abstract Validator class inherit from an interface:

```
    public interface IValidator
    {
        bool IsValid(object value);
        string GetMessage();
    }
```

And then we can make the ValidatorCollection in herit from the same interface.  So essentially you could pass an `IValidator` to someone and they could use it and it could be a single `Validator` or a `ValidatorCollection` and they wouldn't need to know or care.  Also, by inheriting from a collection already we wouldn't need to hold a collection as a private variable.

Use `List<>` will guarantee ordering of how they were added so that shuold work fine.  I think this class could look like this:

```
    public class ValidatorCollection : List<Validator>, IValidator
    {
        private string _message;

        public string GetMessage()
        {
            return _message;
        }

        public bool IsValid(object value)
        {
            var isValid = this.All(x => x.IsValid(value));

            if (!isValid)
            {
                _message = this.Where(x => !x.IsValid(value))
                    .Select(x => x.GetMessage())
                    .Aggregate((i, j) => i + j);
            }
            else
            {
                _message = null;
            }
            

            return isValid;
        }
    }
```

hbgwhite replied:
Looks great. I'll refactor

jtreuting commented on TabulateSmarterTestAdminPackage.Common/Validators/ValidatorCollection.cs:
```diff
@@ -0,0 +1,55 @@
+            return _validators
+                .Where(x => !x.IsValid(value))
+                .Select(x => x.GetMessage())
+                .Aggregate((i, j) => i + j);
```
Do we need to aggregate the validation messages?  For example, with a Required String check and a Minimum Length check (maybe not a realistic example, but to show a point) would we want them both there or is it enough to return the first one that fails?

If we need all of them, I don't think this Aggregate would comma separate them.  I think you'd want:

`String.Join(", ", this.Where(x => !x.IsValid(value)).Select(x => x.GetMessage()).ToArray());`

hbgwhite replied:
I was going to throw all of the different errors in brackets "[Error1][Error2] and return them all as a single row per item that's failing validation. I don't think there's any harm in returning all the things wrong with a particular field. The format would look like: $"{path} attribute {attribute} violates {violationMessage}" (AdminPackageUtility class)

jtreuting commented on TabulateSmarterTestAdminPackage/Processors/Loader/TestSpecificationErrorProcessor.cs:
```diff
@@ -0,0 +1,36 @@
+            {
+                new RequiredStringValidator(ErrorSeverity.Severe, null),
+                new MaxLengthValidator(ErrorSeverity.Severe, 100)
+            });
```
If you switch to the `ValidatorCollection` that inherits from `List<Validator>` then you can instantiate these like:

```
new ValidatorCollection() {
                new RequiredStringValidator(ErrorSeverity.Severe, null),
                new MaxLengthValidator(ErrorSeverity.Severe, 100)
            };
```

jtreuting commented on TabulateSmarterTestAdminPackage/Utility/AdminPackageUtility.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            if (ItemWriter != null)
+            {
+#if DEBUG
```
don't need this because `Debug.Fail` handles the conditional compilation

jtreuting reviewed (changes requested):
Overall this is looking really good.  I have some thoughts for changes.  Let me know if you want to discuss any of them offline

hbgwhite committed:
Addressing JT's pull request comments:
 - Lots of formatting changes
 - IValidator interface
 - ValidatorCollection extends List<Validator> and implements IValidator now
 - All validators currently in project now implement IValidator
 - All processors modified to use new validator syntax
 - Unnecessary intermediate steps in casting -> validation removed
 - Extra #DEBUG statements in code removed
 - Parameter to validators that don't require a parameter has been made null by default and calling processors updated to reflect this change

hbgwhite committed:
- Updated TestSpecificationProcessor IsExpectedPackagePurpose to use new Validator syntax
- Added StringMatchValidator
- Added PoolPropertyProcessor

hbgwhite committed:
- AdminSegmentProcessor stub
- AdministrationProcessor
- ItemPoolProcessor stub
- PoolPropertyProcessor
- TestBlueprintProcessor
- Modified TestSpecificationProcessor to include AdministrationProcessor property
- BlueprintElementProcessor stub

hbgwhite committed:
- BlueprintElementProcessor
- BlueprintIdentifierProcessor

hbgwhite committed:
- ItemPoolProcessor
- FilePathValidator
- PassageIdentifierProcessor
- PassageProcessor
- TestItemProcessor stub

hbgwhite committed:
Formatting

hbgwhite committed:
- TestItemProcessor stub
- BPrefProcessor stub
- ItemScoreParameterProcessor stub
- ItemScoredDimension stub
- TestItemIdentifierProcessor
- TestItemPoolPropertyProcessor

hbgwhite committed:
- ItemScoredDimensionProcessor
